### ISSUE TITLE
Allow an interface to be properly used as a structured generic type argument

### ISSUE THREAD

**Issue by soul-codes** [microsoft/TypeScript#37336]
## Search Terms

interface extends generic argument

### Currently
There is a [difference](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=11&pc=2#code/JYOwLgpgTgZghgYwgAgGIHt0AU5TgW2QG8BYAKGWQCFcAuZdAIwCsIEwBucgX3PJgCuIdsHQhkMTAB4AKsggAPSCAAmAZzSYcefAD4AFAEpi5SgjFqwyANb0ZAbQBENKI4C6yALzFuHZAHp-ZAB5AGkePjJBYTBRcUY4AC9ZeSUIVQ0mVnYDY1IKZHMQSxs7Lx8-QPkoKHQoHiA) between the validity of assignment to `k` in `foo` and `baz`. 

```ts
interface FooParam {
  Bar: object;
}

function foo<T extends FooParam>() {
  const k: T["Bar"] = {}; // OK
}

function baz<T extends object>() {
  const k: T = {}; // error
}
```

In `baz`, it is not possible to assign the empty object to `k` even though `k` is type `T` and `T extends object`. This, as I understand, is the correct behavior since an empty object cannot be guaranteed to be a subtype of `T` even though it is a subtype of `object`. (For instance, `T` could be `number[]` and then we'd have a problem).

In `foo`, on the other hand, it is possible to assign the empty object to `k`, even though it cannot be guaranteed that the empty object would be a subtype of `T["Bar"]`. For example, `T` could be something like `{ Bar: number[] }`, and we would have essentially the same problem as in the case of `baz`.

### Wanted

It should not be possible to assign to `k` in both cases.

## Use Cases

I am currently using the above strategy of submitting an `interface` as generic argument as to reduce the number of places of arguments in the generics and make it more resistant to semantic changes in order. 

The situation that I am facing is that of a functionality that needs many delegates that accepting a large set of different kinds of inputs and returning different kinds of outputs, but the exact set of inputs and outputs are kept generic and type-checked. Using placed generic arguments is error-prone and hard to read.

As it is currently, protection against the type not even being a subtype of the interface property type is already afforded (so, with the example above, assigning `2` to `k` which expects `T["Bar"]` is an error). This issue just requests that the full protection afforded by "standalone" type parameters are also afforded here.

## Examples

```ts
function foo<T extends /*...*/,U  extends /*...*/,V /*... and so on */,W,X,Y,Z>(input: T,
   logic1: (input: T) => U,
   logic2: (input: U) => W,
   logic3: (input: W) => X,
   logic4: (input1: W, input2: X) => U,
   /* ... many more contextual callbacks */
) {
   /* ... */
}
```

This would then turn to something like so:

```ts
interface FooParams {
   T: /* ... whatever T extends */
   U: /* ... whatever U extends */,
   /* ... and so on, obviously with more sensible names */
}

function foo<Params extends FooParams>(input: Params["T"],
   logic1: (input: Params["T"]) => Params["U"],
   /* ... */
) {
   /* ... */
}
```

## Checklist

My suggestion meets these guidelines:

* :question:  This wouldn't be a breaking change in existing TypeScript/JavaScript code. 

:point_up: JavaScript wouldn't break obviously. But since the TypeScript behavior would restrict what are currently valid typings, I'm honestly not sure because I don't know how idiomatic this usage is elsewhere and what/whether people currently actively expect to be the behavior.

* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by MartinJohns** [comment]
> In foo, on the other hand, it is possible to assign the empty object to k, even though it cannot be guaranteed that the empty object would be a subtype of T["Bar"]. For example, T could be something like { Bar: number[] }, and we would have essentially the same problem as in the case of baz.

That's not right. The generic constraint specifies that `T` extends `FooParam`, and `FooParam['Bar']` has the type `object`. This is something that every type extending `FooParam` has to follow, so it's safe to assign an empty object. The type is always `object`.

In your `baz` example however the type is not always `object`, but it can be a type extending `object` that provides additional required properties - so it's not safe to assign an empty object.

**Comment by soul-codes** [comment]
> FooParam['Bar'] has the type object. This is something that every type extending FooParam has to follow, **so it's safe to assign an empty object. The type is always object.**

Hm, I'm pretty sure this isn't right either. Yes, a type extending `FooParam` must have the property `Bar` being an `object`, **but the property `Bar`'s type could also be a subtype of `object`**; i.e. in this situation, any type that can be assigned to a place expecting `object` (for instance, `string[]` is also an `object`). The emphasized quote seems to have not accounted for the subtype scenario.

To demontrate that this property-wise treatment of `extends` is what's happening, and why it is therefore **not** safe to assume that the type is always object, consider the code below.

Here, `ArrayBar` has a a property `Bar` perfectly capable of being assigned into `object`, as `string[]` can be assigned into `object`. Therefore `ArrayBar` extends `FooParam`. The snippet shows that the generic argument typecheck will pass but the `object`-and-only-`object` assumption now [allows an object to be returned in a place expecting an array](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgGIHt0AU5TgW2QG8BYAKGWQCFcAuZdAIwCsIEwBucgX3PNEixEKAIJQ8ATxpRi5StPoBnMFFABzANoBdHlzLkE6EMuT4JYyfQtwpuZAF5i1Osg0AiABYQANt-RutZG49cgB6UOQwD2BFZDhFRWA1Y2QAeQBpABpIrygIGHQ803NxG2QYuORFAFdGMAkABxR0GDRMHDx8ADowiOg1dGRrWxkKuCra+qaGVoxsXAJe5AgANwgQctbpctiQdDBlgA9EMG8JBhY2MB6yQ2MD5WBfAFkSyzb5zodi4Y5kPjIMGqIHYwCMyAK6AAPAAVI6QEAAE1icw6BAAfAAKACU9Bh7mkAVkFGQeTA1SgGyIwWQ4WQewOcA2uEk2UUgwQTIA5AcGtVFB5spyQDzTHAGtkIGAEDdePoyHSAOooRRwYCIoalEY7So1OqNZqzdoLfDZVRqDxgAD85EhUOG0ix2K6+HFmOAkEI9nRyDu7O8EC6fjUmLc1Q8DA8bmx2I4QA).

```ts
interface FooParam {
  Bar: object;
}

interface ArrayBar {
  Bar: string[]
};

const myArray: ArrayBar = { Bar: ["hello"] };

// this assigns OK, therefore myArray is a subtype of FooParam.
// ergo ArrayBar is a subtype of FooParam
// even if Bar is not exactly object.
const stillMyArray: FooParam = myArray; 

function foo<T extends FooParam>(): T["Bar"] {
  return {}; // not an array, so can't push, can't map, etc.
}

// We said ArrayBar is a subtype of FooParam, right?
foo<ArrayBar>().map(item => console.log("uh oh"));
```

**Comment by RyanCavanaugh** [comment]
@soul-codes is correct that this is a soundness hole, a small repro would be:
```ts
type StringBox = { p: string };
function fn<T extends StringBox>(s: T): T["p"] {
  //     vvvvv should error
  return "foo";
}
const a: "bar" = fn({ p: "bar" } as const);
```
wherein `"foo"` inhabits a `"bar"` variable.
